### PR TITLE
Add the possibility to uninstall conda environments that are not in u…

### PR DESCRIPTION
**Pull request by mvdbeek**
…se by Galaxy.

### CODE REVIEW

**Comment by mvdbeek** [comment]
Alright, this is working pretty well for me (saved me a couple of GB), I'd appreciate a review!

File: lib/galaxy/tools/deps/views.py [hunk 1]
```diff
@@ -71,6 +71,26 @@ def uninstall_dependencies(self, index=None, **payload):
                     return return_code
         return None
 
+    @property
+    def unused_dependency_paths(self):
+        """List dependencies that are not currently installed."""
+        unused_dependencies = []
+        for resolver in self._dependency_resolvers:
+            if hasattr(resolver, 'unused_dependency_paths'):
+                unused_dependencies.extend(resolver.unused_dependency_paths(self.toolbox_requirements_status))
```
Comment (nsoranzo) [inline lib/galaxy/tools/deps/views.py]: You can save `self.toolbox_requirements_status` to a variable before the `for` loop.

File: lib/galaxy/tools/deps/views.py [hunk 2]
```diff
@@ -71,6 +71,26 @@ def uninstall_dependencies(self, index=None, **payload):
...
+        return set(unused_dependencies)
+
+    def remove_unused_dependency_paths(self, envs):
+        """Remove dependencies that are not currently used."""
+        envs_to_remove = set(envs)
+        for resolver in self._dependency_resolvers:
+            if hasattr(resolver, 'unused_dependency_paths') and hasattr(resolver, 'uninstall_environments'):
+                unused_dependencies = resolver.unused_dependency_paths(self.toolbox_requirements_status)
```
Comment (nsoranzo) [inline lib/galaxy/tools/deps/views.py]: Same as above.

File: lib/galaxy/tools/deps/views.py [hunk 3]
```diff
@@ -71,6 +71,26 @@ def uninstall_dependencies(self, index=None, **payload):
...
+                can_remove = envs_to_remove & set(unused_dependencies)
+                exit_code = resolver.uninstall_environments(can_remove)
+                if exit_code == 0:
+                    envs_to_remove = envs_to_remove.difference(can_remove)
```
Comment (nsoranzo) [inline lib/galaxy/tools/deps/views.py]: Should this method return something?

**Comment by jmchilton** [comment]
Seems like a solid idea to me and the implementation looks good! Thanks for your continued work on this GUI @mvdbeek - this is quite exciting and most appreciated. 👍

**Comment by dannon** [comment]
Looks nice, and it worked to clean up a couple things for me!

**Pull request merged**